CorsixTH/Src/th_lua.cpp

```diff
@@ -320,58 +320,47 @@ void luaT_pushtablebool(lua_State* L, const char* k, bool v) {
   lua_settable(L, -3);
 }
 
+void luaT_printvalue(lua_State* L, int idx) {
+  int t = lua_type(L, idx);
+  switch (t) {
+    case LUA_TSTRING: /* strings */
+      std::printf("string: '%s'\n", lua_tostring(L, idx));
+      break;
+    case LUA_TBOOLEAN: /* booleans */
+      std::printf("boolean %s\n", lua_toboolean(L, idx) ? "true" : "false");
+      break;
+    case LUA_TNUMBER: /* numbers */
+      std::printf("number: %g\n", lua_tonumber(L, idx));
+      break;
+    default: /* other values */
+      std::printf("%s: %p\n", lua_typename(L, t), lua_topointer(L, idx));
+      break;
+  }
+}
+
 void luaT_printstack(lua_State* L) {
   int i;
   int top = lua_gettop(L);
 
   std::printf("total items in stack %d\n", top);
 
   for (i = 1; i <= top; i++) { /* repeat for each level */
-    int t = lua_type(L, i);
-    switch (t) {
-      case LUA_TSTRING: /* strings */
-        std::printf("string: '%s'\n", lua_tostring(L, i));
-        break;
-      case LUA_TBOOLEAN: /* booleans */
-        std::printf("boolean %s\n", lua_toboolean(L, i) ? "true" : "false");
-        break;
-      case LUA_TNUMBER: /* numbers */
-        std::printf("number: %g\n", lua_tonumber(L, i));
-        break;
-      default: /* other values */
-        std::printf("%s\n", lua_typename(L, t));
-        break;
-    }
-    std::printf(" "); /* put a separator */
+    printf("(%d) ", i);
+    luaT_printvalue(L, i);
   }
   std::printf("\n"); /* end the listing */
 }
 
 void luaT_printrawtable(lua_State* L, int idx) {
-  int i;
-  int len = static_cast<int>(lua_objlen(L, idx));
-
-  std::printf("total items in table %d\n", len);
-
-  for (i = 1; i <= len; i++) {
-    lua_rawgeti(L, idx, i);
-    int t = lua_type(L, -1);
-    switch (t) {
-      case LUA_TSTRING: /* strings */
-        std::printf("string: '%s'\n", lua_tostring(L, -1));
-        break;
-      case LUA_TBOOLEAN: /* booleans */
-        std::printf("boolean %s\n", lua_toboolean(L, -1) ? "true" : "false");
-        break;
-      case LUA_TNUMBER: /* numbers */
-        std::printf("number: %g\n", lua_tonumber(L, -1));
-        break;
-      default: /* other values */
-        std::printf("%s\n", lua_typename(L, t));
-        break;
-    }
-    std::printf(" "); /* put a separator */
+  /* table is in the stack at index 't' */
+  lua_pushnil(L); /* first key */
+  while (lua_next(L, idx) != 0) {
+    /* uses 'key' (at index -2) and 'value' (at index -1) */
+    printf("key: ");
+    luaT_printvalue(L, -2);
+    printf("value: ");
```

Comment:
and more `printf`, at the left we have `std::printf`

Reply:
Good catch, seems every compiler I use wants to import c functions into the global scope.